vespa-http-client/src/main/java/com/yahoo/vespa/http/client/core/communication/IOThread.java

```diff
@@ -348,11 +347,16 @@ private ConnectionState cycle(ConnectionState connectionState) {
                     currentConnection.handshake();
                     successfulHandshakes.getAndIncrement();
                 } catch (ServerResponseException ser) {
+                    int code = ser.getResponseCode();
+                    if (code == 401 || code == 403) {
+                        drainDocumentQueueWhenFailingPermanently(new Exception("Denied access by endpoint:" + ser.getResponseString()));
+                        log.log(Level.SEVERE, "Failed authentication or authorization with '" + endpoint + "': " + Exceptions.toMessageString(ser));
+                        return ConnectionState.CONNECTED; // Should ideally exit immediately, instead of doing this per X documents :/
```

Comment:
If we want to we could exit, but I'm not sure it's always the best thing to do since there could be a fix to the problem rolling out simultaneously, we could be feeding to multiple clusters etc. I think in general this client wants to keep trying until timeout, for better or worse.

Reply:
Yep.